VSD - liveness context

jezhiggins opened the pull request:
Adds a new abstract_objectt in the `context_abstract_objectt` hierarchy which stores a single location, indicating where this value became live.

- **Transform** : if the instruction is an assignment and a variable is written to (including merge 'maybe' writes) then the location is set to the current instruction.

- **Merge** : if the instruction is the same on both sides of the merge, then use that value. If it is different then use the location of the merge point.

jezhiggins committed:
Initial work to introduce new region_context

At the moment, the region_context is identical in operation to
write_location_context. The primary purpose of this initial commit is i
add the necessary config options and factory changes.

jezhiggins committed:
Skeleton region_context implementation.

Holds last write location. Clears location on merge.

jezhiggins committed:
Bring full_array_ and full_struct_ao visit_sub_element implementations in line with each other

jezhiggins committed:
Pull out map_visit

jezhiggins committed:
Pull location_update_visitort out into its own header file.

Correct abstract_object_visitort::visit to take a const-ref parameter
rather than by-value. (Since we're passing a std::shared_ptr it's not a
hugely significant change, but I'm sure that was what was originally
intended.)

jezhiggins committed:
Fill out full_*_abstract_object::update_location_context

Make update_location_context complete in the relevant object, rather
than needing the caller to then call visit_sub_elements.

jezhiggins committed:
Remove unused data_dependency_contextt::insert_data_deps(locationst)

jezhiggins committed:
Rework context_abstract_object::update_location_context

Move common function down into context_abstract_object. Add
update_location_context_internal where derived classes implement their
specific needs. Prevents calling down through the base classes, and
then having to dynamic_pointer_cast the result. Clearer all around.

jezhiggins committed:
Add abstract_object::write_location_context

Default implementation just forwards to update_location_context({from}, true)

jezhiggins committed:
Simplify update_location_context

Eliminate update_sub_elements flag parameter as it would always be true.

jezhiggins committed:
Consolidate write_location_context, eliminate update_location_context

jezhiggins committed:
Add merge_location parameter to abstract_environment::merge

jezhiggins committed:
Pass merge_location to newly merged object

Extend region-contextt to pick this up if it needs to use it. Ever other
object just ignores it.

jezhiggins committed:
Change region_contextt to liveness_contextt

jezhiggins committed:
Further liveness regression tests

jezhiggins committed:
Propagate merge location context down to struct members

jezhiggins committed:
Propagate merge location context down to array elements

martin-cs commented on src/analyses/variable-sensitivity/variable_sensitivity_domain.h:
```diff
@@ -78,8 +78,8 @@
-  "(vsd-data-dependencies)"
-
+  "(vsd-data-dependencies)"                                                    \
+  "(vsd-region-tracking)"
```
A backslash and a blank new line after so that future additions are just one line?

martin-cs commented on src/analyses/variable-sensitivity/region_context.cpp:
```diff
@@ -0,0 +1,373 @@
+ * \param out the stream on which to output the locations
+ * \param locations the set of locations to output
+ */
+void region_contextt::output_last_written_locations(
```
Is this needed?  If it is general to all of the domains that track locations then maybe it could be put somewhere more general.  If not then could it be simplified as only one location is stored.

martin-cs commented on regression/goto-analyzer/liveness-struct/test-write-location.desc:
```diff
@@ -0,0 +1,9 @@
+CORE
+main.c
+--variable-sensitivity --vsd-structs every-field --vsd-values intervals --vsd-liveness --show
```
Should this have `--vsd-liveness` set?

jezhiggins replied:
nope!

martin-cs commented on regression/goto-analyzer/liveness/test-liveness.desc:
```diff
@@ -0,0 +1,9 @@
+^SIGNAL=0$
+^main::1::x .* TOP @ \[1\]
+^main::1::x .* value-set-begin: 0 :value-set-end @ \[2\]
+^main::1::x .* value-set-begin: 2, 3 :value-set-end @ \[7\]
```
Pedantry probably but could we have the other two regions for illustration and completeness.

martin-cs commented on src/analyses/variable-sensitivity/abstract_object.cpp:
```diff
@@ -206,6 +206,19 @@ void abstract_objectt::output(
+  const widen_modet &widen_mode)
+{
+  abstract_object_pointert result = merge(op1, op2, widen_mode).object;
+  result = result->merge_location_context(merge_location);
```
Is a separate call better than an extra argument here?

martin-cs replied:
Ah; I think I see it.  It is the partner of `write_location_context` which can't easily be folded into eval.

martin-cs commented on src/analyses/variable-sensitivity/liveness_context.cpp:
```diff
@@ -0,0 +1,220 @@
+  abstract_object_pointert updated_child = child_abstract_object->write(
+    environment, ns, stack, specifier, value, merging_write);
+
+  // Only perform an update if the write to the child has in fact changed it...
```
I'm not exactly sure what (code level) condition this is checking?

martin-cs commented on src/analyses/variable-sensitivity/liveness_context.cpp:
```diff
@@ -0,0 +1,220 @@
+  // For two sets of last written locations to match,
+  // each location in one set must be equal to precisely one location
+  // in the other, since a set can assume at most one match
+  return before_context->get_location() != get_location();
```
As `locationt`'s are iterators into (potentially different) lists, we should compare location numbers instead of directly comparing locations.

martin-cs commented on src/analyses/variable-sensitivity/variable_sensitivity_domain.h:
```diff
@@ -78,8 +78,8 @@
-  "(vsd-data-dependencies)"
-
+  "(vsd-data-dependencies)"                                                    \
+  "(vsd-liveness)"
```
Add a backslash and a new line so that future additions are single lines?

martin-cs reviewed (approved):
Thanks for the tidy up and refactoring as well as the new abstract object.  Please can you add a test case involving function calls.  It needs a function that is called at least twice and a version with and without `--three-way-merge`.

jezhiggins committed:
liveness fixes for three-way-merge

jezhiggins committed:
Rework liveness_contextt on top of write_location_contextt

Maintains existing three-way-merge behaviour, which otherwise was lost
because we'd discarded too much context information to be able to merge
precisely.

martin-cs reviewed (approved):
Thanks for chasing up the three-way-merge issues.
